### PR TITLE
Rolling Update Strategy

### PR DESCRIPTION
**Pull request by reset**
This is the MVP candidate for adding a rolling update strategy to the supervisor via the `--strategy rolling` flag. The rolling update strategy behaves as follows:

* An update leader is elected. If service group was started with the leader topology the leader will not be elected as the update leader.
* The update leader will wait to check for new versions until all services within the service group agree on their version.
* Members of the service group will look at their "previous peer" to determine if it is their turn to update.
* Services are updated one at a time.
* Updates are eventually consistent. If a service joins the service group late it may update out of order, but it will be on the right version.

gif-keyboard-9170833510054505331

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/butterfly/protocols/swim.proto [hunk 1]
```diff
@@ -51,6 +51,7 @@ message Service {
   optional uint32 port = 6;
   repeated uint32 exposes = 7;
   optional bool initialized = 8;
+  optional string package_ident = 9;
```
Comment (adamhjk) [inline components/butterfly/protocols/swim.proto]: Of course this should have been here.

File: components/butterfly/src/rumor/election.rs [hunk 2]
```diff
@@ -210,6 +212,76 @@ impl Rumor for Election {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct ElectionUpdate(Election);
```
Comment (adamhjk) [inline components/butterfly/src/rumor/election.rs]: I'm going to name using structs with one member this way the Jamie Struct.

File: components/butterfly/src/rumor/service.rs [hunk 3]
```diff
@@ -80,16 +81,23 @@ impl DerefMut for Service {
...
+                                   exposes: Vec<u32>)
+                                   -> Self
         where S1: Into<String>,
               S2: Into<String>,
-              S3: Into<String>
+              S3: Into<String>,
+              S4: Identifiable,
+              S5: Into<String>
     {
+        assert!(package.fully_qualified(),
```
Comment (adamhjk) [inline components/butterfly/src/rumor/service.rs]: I don't think there is anything to do about it right this second, but I have a lurking fear that we're taking information from the network and possibly causing ourselves to panic. I think that's not the case here, but I can feel it in Butterfly.

File: components/butterfly/src/server/mod.rs [hunk 4]
```diff
@@ -438,26 +439,40 @@ impl Server {
...
+        let mut update_elections_to_restart = vec![];
+
         self.election_store.with_keys(|(service_group, rumors)| {
             if self.service_store.contains_rumor(&service_group, self.member_id()) {
                 // This is safe; there is only one id for an election, and it is "election"
                 let election = rumors.get("election")
                     .expect("Lost an election struct between looking it up and reading it.");
                 // If we are finished, and the leader is dead, we should restart the election
-                if election.get_status() == Election_Status::Finished && election.get_member_id() == self.member_id() {
+                if election.is_finished() && election.get_member_id() == self.member_id() {
```
Comment (adamhjk) [inline components/butterfly/src/server/mod.rs]: nice cleanup

File: components/core/src/service.rs [hunk 5]
```diff
@@ -50,6 +50,8 @@ impl ServiceGroup {
     }
 
     pub fn as_string(&self) -> String {
+        // JW TODO: We shouldn't allocate a new string each time we call this. We should store the
```
Comment (adamhjk) [inline components/core/src/service.rs]: so much this.

**Comment by adamhjk** [comment]
gif-keyboard-1180692586422397142

**Comment by adamhjk** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/185937418).

File: components/butterfly/src/rumor/service.rs (hunk 3 above)
Comment (reset) [inline components/butterfly/src/rumor/service.rs]: Yeah since this is the network layer we might want to just drop the rumor instead of a panic

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/185937418).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**